### DISCUSSION TITLE
Is there a component for turning an mdast Root into React Elements?

### DISCUSSION THREAD

**Discussion by FunctionDJ** [remarkjs/remark#971] [Q&A] [answered]
My site (Next.js) renders a bunch of Markdown files that won't change at runtime, so i found this article: https://css-tricks.com/responsible-markdown-in-next-js/
I've already coded all the node types i need and also improved on this guide with good TypeScript typings, but i was wondering if there's already a solution out there that doesn't require too much hacking.

This is what i've made so far:
```tsx
import { memo } from 'react';
import { Content } from "mdast";
import { ParseableNode } from './parse-markdown';

const children = (node: { children: Content[] }) => ( // TODO memoize
  node.children.map((n, i) => <MarkdownNode key={i} node={n}/>)
);

const MarkdownNode = ({ node }: { node: ParseableNode }): JSX.Element => {
  switch (node.type) {
    case "root": return <>{children(node)}</>;
    case "paragraph": return <p>{children(node)}</p>;
    case "emphasis": return <em>{children(node)}</em>;
    case "link": return <a title={node.title ?? undefined} href={node.url}>{children(node)}</a>;
    case "heading": {
      const Heading = `h${node.depth}` as const;
      return <Heading>{children(node)}</Heading>;
    }
    case "text": return <>{node.value}</>;
    case "break": return <br/>;
    case "strong": return <strong>{children(node)}</strong>;
    case "inlineCode": return <code>{node.value}</code>;
    case "list": return node.ordered ? <ol>{children(node)}</ol> : <ul>{children(node)}</ul>;
    case "listItem": return <li>{children(node)}</li>;
  }

  return <em>Unsupported Node: {node.type}</em>;
};

export const MarkdownRenderer = memo(MarkdownNode);
```

**Comment by ChristianMurphy** [comment, marked as answer]
There are a few options depending on how you want to handle it:
https://github.com/remarkjs/react-remark#server-side-rendering
https://mdxjs.com/docs/getting-started/#nextjs

are two notable ones.
One ties into next's static props feature (more on this in https://github.com/remarkjs/react-remark/issues/16)
The other ties into next's webpack build

**Reply by FunctionDJ** [reply]
I need to be able to customize the rendering behaviour in the client because i'm customizing `<h4>` rendering, so the first one sadly isn't an option.
Regarding MDX, i've had this in my code since i started this:
```js
  // TODO instead of this custom stuff i should probably use MDX instead
```
😄 Luckily what i have right now seems to be working pretty well and i don't need to ship a markdown parser to clients.
Still, thanks a lot for your response!

**Reply by ChristianMurphy** [reply]
> I need to be able to customize the rendering behaviour in the client because i'm customizing `<h4>` rendering, so the first one sadly isn't an option.

I'm not sure I follow, https://github.com/remarkjs/react-remark#options includes a `components` sub-option (https://github.com/rehypejs/rehype-react#optionscomponents) which does allow components to be re-mapped / customized

**Reply by FunctionDJ** [reply]
@ChristianMurphy as far as i can see, those options would still ship remark in the browser, which i'd like to avoid because my content doesn't change. maybe the article i linked helps to illustrate my question: https://css-tricks.com/responsible-markdown-in-next-js/

**Reply by wooorm** [reply]
I believe you are overthinking the problem. There are several existing solutions:
- https://mdxjs.com/guides/mdx-on-demand/
- https://github.com/hashicorp/next-mdx-remote
- https://github.com/kentcdodds/mdx-bundler
- Remix by default
...and more will come with React Server Components.

Additionally, if you content doesn’t change, and you trust it (note that you can make it save on the server), just send a string of HTML down the wire.

**Reply by FunctionDJ** [reply]
@wooorm Those are great resources! I guess i should just use dangerouslySetInnerHTML like you said. It makes things a lot simpler and i need a lot less code without sacrificing anything really. Thank you!
